src/Shared/PlatformNegotiation.cs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.BackEnd;
+
+namespace Microsoft.Build.Shared
+{
+    /// <summary>
+    /// This class contains only static methods, which are useful throughout many
+    /// of the XMake classes and don't really belong in any specific class.   
+    /// </summary>
+    static internal class PlatformNegotiation
+    {
+        internal static string? GetPlatformString(String projectReferencePlatformMetadata, String projectReferenceLookupTableMetadata, String CurrentProjectPlatformMetadata, String CurrentPlatformLookupTableMetadata, String ProjectPath, TaskLoggingHelper? Log = null)
+        {
+            
+            Dictionary<string, string>? currentProjectLookupTable = ExtractLookupTable(CurrentPlatformLookupTableMetadata, Log);
+
+            if (string.IsNullOrEmpty(projectReferencePlatformMetadata))
+            {
+                if(Log != null)
+                {
+                    Log.LogWarningWithCodeFromResources("GetCompatiblePlatform.NoPlatformsListed", ProjectPath);
+                }
+                return null;
+            }
+            Dictionary<string, string>? projectReferenceLookupTable = ExtractLookupTable(projectReferenceLookupTableMetadata, Log);
+            HashSet<string> projectReferencePlatforms = new HashSet<string>();
+            foreach (string s in projectReferencePlatformMetadata.Split(MSBuildConstants.SemicolonChar, StringSplitOptions.RemoveEmptyEntries))
+            {
+                projectReferencePlatforms.Add(s);
+            }
+
+            string buildProjectReferenceAs = string.Empty;
+
+            if (projectReferencePlatforms.Contains(CurrentProjectPlatformMetadata))
+            {
+                buildProjectReferenceAs = CurrentProjectPlatformMetadata;
+                if (Log != null)
+                {
+                    Log.LogMessageFromResources(MessageImportance.Low, "GetCompatiblePlatform.SamePlatform");
+                }
+            }
+            // Prioritize PlatformLookupTable **metadata** attached to the ProjectReference item
+            // before the current project's table. We do this to allow per-ProjectReference fine tuning.
+            else if (projectReferenceLookupTable != null &&
+                    projectReferenceLookupTable.ContainsKey(CurrentProjectPlatformMetadata) &&
+                    projectReferencePlatforms.Contains(projectReferenceLookupTable[CurrentProjectPlatformMetadata]))
```

Comment:
Can avoid double lookup here.

```suggestion
                    projectReferenceLookupTable.TryGetValue(CurrentProjectPlatformMetadata, out var projectReference) &&
                    projectReferencePlatforms.Contains(projectReference))
```